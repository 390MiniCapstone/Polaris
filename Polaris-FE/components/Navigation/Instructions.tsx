import React from 'react';
import { View, Text, StyleSheet } from 'react-native';

interface InstructionsProps {
  instruction: string;
}

export const Instructions: React.FC<InstructionsProps> = ({ instruction }) => {
  return (
    <View style={styles.container}>
      <Text style={styles.instruction}>{instruction}</Text>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    position: 'absolute',
    top: 64,
    left: 15,
    right: 15,
    backgroundColor: 'rgba(34, 34, 34, 0.992)',
    borderRadius: 20,
    paddingHorizontal: 20,
    paddingVertical: 10,
    height: 80,
    justifyContent: 'center',
  },
  instruction: {
    color: 'white',
    fontSize: 20,
    fontWeight: '500',
    textAlign: 'left',
  },
});
