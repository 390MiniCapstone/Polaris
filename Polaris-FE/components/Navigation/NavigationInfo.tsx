import React from 'react';
import { View, Text, StyleSheet, TouchableOpacity } from 'react-native';
import { FontAwesome5 } from '@expo/vector-icons';
import { handleCurrentLocation } from '@/utils/mapHandlers';
import { openTransitInMaps } from '@/utils/navigationUtils';
import { mapRef } from '@/utils/refs';
import { useMapLocation } from '@/hooks/useMapLocation';

interface NavigationInfoProps {
  duration: number;
  distance: number;
  isNavigating: boolean;
  is3d: boolean;
  travelMode: string;
  destination: { latitude: number; longitude: number };
  onCancel: () => void;
  onStartNavigation: () => void;
  onCurrentLocation?: () => void;
  updateIs3d?: (value: ((prevState: boolean) => boolean) | boolean) => void;
}

export const NavigationInfo: React.FC<NavigationInfoProps> = ({
  duration,
  distance,
  isNavigating,
  onCancel,
  onStartNavigation,
  is3d,
  travelMode,
  updateIs3d,
  destination,
}) => {
  const { location } = useMapLocation();

  return (
    <View style={styles.container}>
      <TouchableOpacity
        testID="cancel-button"
        style={styles.cancelButton}
        onPress={onCancel}
      >
        <FontAwesome5 name="times" size={20} color="white" />
      </TouchableOpacity>

      <View style={styles.infoContainer}>
        <Text style={styles.timeText}>{Math.ceil(duration / 60)} Minutes</Text>
        <Text style={styles.timeText}>Â·</Text>
        <Text style={styles.distanceText}>
          {(distance / 1000).toFixed(1)} km
        </Text>
      </View>

      <TouchableOpacity
        testID="action-button"
        style={isNavigating ? styles.currentButton : styles.goButton}
        onPress={
          isNavigating
            ? () => {
                if (is3d) {
                  handleCurrentLocation(mapRef, location);
                  updateIs3d && updateIs3d(false);
                } else {
                  onStartNavigation();
                  updateIs3d && updateIs3d(true);
                }
              }
            : () => {
                if (travelMode === 'TRANSIT' && location) {
                  openTransitInMaps(location, destination);
                } else {
                  onStartNavigation();
                }
              }
        }
      >
        {isNavigating ? (
          <FontAwesome5 name="location-arrow" size={16} color="white" />
        ) : (
          <Text style={styles.goButtonText}>GO</Text>
        )}
      </TouchableOpacity>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    position: 'absolute',
    bottom: 75,
    left: 20,
    right: 20,
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
  },
  cancelButton: {
    backgroundColor: 'rgba(34, 34, 34, 0.992)',
    borderRadius: 12,
    height: 36,
    width: 52,
    justifyContent: 'center',
    alignItems: 'center',
  },
  infoContainer: {
    backgroundColor: 'rgba(34, 34, 34, 0.992)',
    paddingHorizontal: 20,
    borderRadius: 16,
    minWidth: 200,
    height: 46,
    flexDirection: 'row',
    justifyContent: 'space-between',
    gap: 5,
    alignItems: 'center',
  },
  timeText: {
    fontSize: 18,
    fontWeight: '600',
    color: 'white',
  },
  distanceText: {
    fontSize: 18,
    color: 'rgba(255, 255, 255, 0.8)',
  },
  currentButton: {
    backgroundColor: 'rgba(34, 34, 34, 0.992)',
    borderRadius: 12,
    height: 36,
    width: 52,
    justifyContent: 'center',
    alignItems: 'center',
  },
  goButton: {
    backgroundColor: '#9A2E3F',
    borderRadius: 12,
    height: 36,
    width: 52,
    justifyContent: 'center',
    alignItems: 'center',
  },
  goButtonText: {
    color: 'white',
    fontSize: 16,
    fontWeight: '600',
  },
});
