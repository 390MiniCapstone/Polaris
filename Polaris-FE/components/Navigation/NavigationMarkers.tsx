import React from 'react';
import { LatLng, Marker } from 'react-native-maps';
import { useNavigation } from '@/contexts/NavigationContext/NavigationContext';
import { FontAwesome6 } from '@expo/vector-icons';
import { TravelMode, ShuttleBusStop } from '@/constants/types';
import { StyleSheet, Text, View } from 'react-native';
import useTheme from '@/hooks/useTheme';

const Buses: React.FC<{
  busPoints: LatLng[] | undefined;
  navigationState: string;
  nextDeparture: string | null;
}> = ({ busPoints, navigationState, nextDeparture }) => {
  if (!busPoints || navigationState === 'default' || !nextDeparture) {
    return null;
  }

  return (
    <>
      {busPoints.map((point: LatLng) => (
        <Marker key={`${point.latitude}-${point.longitude}`} coordinate={point}>
          <FontAwesome6
            testID="shuttle-marker"
            name="van-shuttle"
            size={18}
            color="white"
          />
        </Marker>
      ))}
    </>
  );
};

const Stops = ({
  travelMode,
  nearestBusStop,
  otherBusStop,
}: {
  travelMode: TravelMode;
  nearestBusStop: ShuttleBusStop | null;
  otherBusStop: ShuttleBusStop | null;
}) => {
  if (travelMode === 'SHUTTLE' && nearestBusStop && otherBusStop) {
    return (
      <>
        <Marker testID="stop-marker" coordinate={nearestBusStop.location}>
          <View style={styles.stopView}>
            <Text style={styles.stopText}>{nearestBusStop.shortName}</Text>
          </View>
        </Marker>
        <Marker testID="stop-marker" coordinate={otherBusStop.location}>
          <View style={styles.stopView}>
            <Text style={styles.stopText}>{otherBusStop.shortName}</Text>
          </View>
        </Marker>
      </>
    );
  }
  return null;
};

export const NavigationMarkers: React.FC = () => {
  const { theme } = useTheme();
  const {
    shuttleData,
    navigationState,
    nextDeparture,
    travelMode,
    nearestBusStop,
    otherBusStop,
    destination,
    clippedPolyline,
    snappedPoint,
  } = useNavigation();

  return (
    <>
      {(navigationState === 'planning' || navigationState === 'navigating') &&
        clippedPolyline &&
        snappedPoint &&
        destination && (
          <Marker
            testID="destination-marker"
            key={theme.colors.primary}
            coordinate={destination}
            pinColor={theme.colors.primary}
          />
        )}

      <Buses
        busPoints={shuttleData?.legTwo?.busPoints}
        navigationState={navigationState}
        nextDeparture={nextDeparture}
      />
      <Stops
        travelMode={travelMode}
        nearestBusStop={nearestBusStop}
        otherBusStop={otherBusStop}
      />
    </>
  );
};

const styles = StyleSheet.create({
  stopView: {
    backgroundColor: '#222',
    flex: 1,
    paddingVertical: 5,
    paddingHorizontal: 10,
    borderRadius: 10,
  },
  stopText: {
    fontSize: 12,
    color: 'white',
    fontWeight: '600',
    textAlign: 'center',
  },
});
