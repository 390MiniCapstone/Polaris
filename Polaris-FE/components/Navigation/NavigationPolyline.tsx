import React from 'react';
import { Marker, Polyline } from 'react-native-maps';
import { useNavigation } from '@/contexts/NavigationContext/NavigationContext';
import useTheme from '@/hooks/useTheme';

export const NavigationPolyline: React.FC = () => {
  const {
    travelMode,
    destination,
    snappedPoint,
    clippedPolyline,
    navigationState,
  } = useNavigation();
  const { theme } = useTheme();

  return (
    <React.Fragment>
      {(navigationState === 'planning' || navigationState === 'navigating') &&
        clippedPolyline &&
        snappedPoint && (
          <React.Fragment>
            <Marker
              key={theme.colors.primary}
              coordinate={destination}
              pinColor={theme.colors.primary}
            />
            {travelMode === 'WALK' ? (
              <Polyline
                coordinates={clippedPolyline}
                strokeWidth={4}
                strokeColor={theme.colors.secondary}
                lineDashPattern={[5, 10]}
              />
            ) : (
              <React.Fragment>
                <Polyline
                  coordinates={clippedPolyline}
                  strokeWidth={9}
                  strokeColor={theme.colors.primary}
                />
                <Polyline
                  coordinates={clippedPolyline}
                  strokeWidth={6}
                  strokeColor={theme.colors.secondary}
                />
              </React.Fragment>
            )}
          </React.Fragment>
        )}
    </React.Fragment>
  );
};
