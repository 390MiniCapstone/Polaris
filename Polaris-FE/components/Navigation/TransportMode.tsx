import React from 'react';
import { View, StyleSheet, TouchableOpacity } from 'react-native';
import { FontAwesome, FontAwesome5, FontAwesome6 } from '@expo/vector-icons';

type TransportMode = 'DRIVE' | 'WALK' | 'TRANSIT' | 'BICYCLE';

interface TransportModeProps {
  selectedMode: TransportMode;
  onModeSelect: (mode: TransportMode) => void;
}

export const TransportMode: React.FC<TransportModeProps> = ({
  selectedMode,
  onModeSelect,
}) => {
  return (
    <View style={styles.container}>
      <TouchableOpacity
        style={[styles.modeButton, selectedMode === 'DRIVE' && styles.selected]}
        onPress={() => onModeSelect('DRIVE')}
        testID="transport-mode-button-DRIVE"
      >
        <FontAwesome6 name="car" size={22} color="white" />
      </TouchableOpacity>
      <TouchableOpacity
        style={[styles.modeButton, selectedMode === 'WALK' && styles.selected]}
        onPress={() => onModeSelect('WALK')}
        testID="transport-mode-button-WALK"
      >
        <FontAwesome5 name="walking" size={22} color="white" />
      </TouchableOpacity>
      <TouchableOpacity
        style={[
          styles.modeButton,
          selectedMode === 'TRANSIT' && styles.selected,
        ]}
        onPress={() => onModeSelect('TRANSIT')}
        testID="transport-mode-button-TRANSIT"
      >
        <FontAwesome6 name="bus-simple" size={22} color="white" />
      </TouchableOpacity>
      <TouchableOpacity
        style={[
          styles.modeButton,
          selectedMode === 'BICYCLE' && styles.selected,
        ]}
        onPress={() => onModeSelect('BICYCLE')}
        testID="transport-mode-button-BICYCLE"
      >
        <FontAwesome name="bicycle" size={22} color="white" />
      </TouchableOpacity>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flexDirection: 'row',
    position: 'absolute',
    top: 76,
    left: 0,
    right: 0,
    justifyContent: 'center',
    gap: 14,
  },
  modeButton: {
    backgroundColor: 'rgba(34, 34, 34, 0.992)',
    borderRadius: 15,
    height: 40,
    width: 60,
    justifyContent: 'center',
    alignItems: 'center',
  },
  selected: {
    backgroundColor: '#9A2E3F',
  },
});
