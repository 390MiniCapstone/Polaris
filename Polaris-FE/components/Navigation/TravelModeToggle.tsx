import React from 'react';
import { View, StyleSheet, TouchableOpacity } from 'react-native';
import { FontAwesome, FontAwesome5, FontAwesome6 } from '@expo/vector-icons';
import { useNavigation } from '@/contexts/NavigationContext/NavigationContext';
import useTheme from '@/hooks/useTheme';

export const TravelModeToggle: React.FC = () => {
  const { theme } = useTheme();

  const { travelMode, setTravelMode } = useNavigation();

  return (
    <View style={styles.container}>
      <TouchableOpacity
        style={[
          styles.modeButton,
          travelMode === 'DRIVE' && { backgroundColor: theme.colors.primary },
        ]}
        onPress={() => setTravelMode('DRIVE')}
        testID="transport-mode-button-DRIVE"
      >
        <FontAwesome6 name="car" size={22} color="white" />
      </TouchableOpacity>
      <TouchableOpacity
        style={[
          styles.modeButton,
          travelMode === 'WALK' && { backgroundColor: theme.colors.primary },
        ]}
        onPress={() => setTravelMode('WALK')}
        testID="transport-mode-button-WALK"
      >
        <FontAwesome5 name="walking" size={22} color="white" />
      </TouchableOpacity>
      <TouchableOpacity
        style={[
          styles.modeButton,
          travelMode === 'TRANSIT' && { backgroundColor: theme.colors.primary },
        ]}
        onPress={() => setTravelMode('TRANSIT')}
        testID="transport-mode-button-TRANSIT"
      >
        <FontAwesome6 name="train-subway" size={22} color="white" />
      </TouchableOpacity>
      <TouchableOpacity
        style={[
          styles.modeButton,
          travelMode === 'BICYCLE' && { backgroundColor: theme.colors.primary },
        ]}
        onPress={() => setTravelMode('BICYCLE')}
        testID="transport-mode-button-BICYCLE"
      >
        <FontAwesome name="bicycle" size={22} color="white" />
      </TouchableOpacity>
      <TouchableOpacity
        style={[
          styles.modeButton,
          travelMode === 'SHUTTLE' && { backgroundColor: theme.colors.primary },
        ]}
        onPress={() => setTravelMode('SHUTTLE')}
        testID="transport-mode-button-SHUTTLE"
      >
        <FontAwesome6 name="van-shuttle" size={22} color="white" />
      </TouchableOpacity>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flexDirection: 'row',
    position: 'absolute',
    top: 76,
    left: 0,
    right: 0,
    justifyContent: 'center',
    gap: 10,
  },
  modeButton: {
    backgroundColor: 'rgba(34, 34, 34, 0.992)',
    borderRadius: 15,
    height: 40,
    width: 60,
    justifyContent: 'center',
    alignItems: 'center',
  },
});
