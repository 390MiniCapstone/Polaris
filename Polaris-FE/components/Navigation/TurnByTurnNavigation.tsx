import React from 'react';
import { View, Text, StyleSheet } from 'react-native';

interface TurnByTurnNavigationProps {
  instruction: string;
}

export const TurnByTurnNavigation: React.FC<TurnByTurnNavigationProps> = ({
  instruction,
}) => {
  return (
    <View style={styles.container}>
      <Text style={styles.instruction}>{instruction}</Text>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    position: 'absolute',
    top: 64,
    left: 15,
    right: 15,
    backgroundColor: 'rgba(34, 34, 34, 0.992)',
    borderRadius: 20,
    paddingHorizontal: 20,
    paddingVertical: 10,
    height: 80,
    justifyContent: 'center',
  },
  instruction: {
    color: 'white',
    fontSize: 20,
    fontWeight: '500',
    textAlign: 'left',
  },
});
