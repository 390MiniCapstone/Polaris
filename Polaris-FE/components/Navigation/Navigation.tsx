import { TouchableOpacity, Text, StyleSheet } from 'react-native';
import { bottomSheetRef, mapRef } from '@/utils/refs';
import React, { SetStateAction, useEffect, useState } from 'react';
import { TravelModeToggle } from '@/components/Navigation/TravelModeToggle';
import { Instructions } from './Instructions';
import { NavigationInfo } from '@/components/Navigation/NavigationInfo';
import { useMapLocation } from '@/hooks/useMapLocation';
import { lineString, point } from '@turf/helpers';
import nearestPointOnLine from '@turf/nearest-point-on-line';
import { handleCurrentLocation } from '@/utils/mapHandlers';
import { NavigationState, TravelMode } from '@/constants/types';
import { LatLng } from 'react-native-maps';
import { RouteData } from '@/constants/types';
import { getGoogleMapsRoute } from '@/services/googleMapsRoutes';
import {
  clipPolylineFromSnappedPoint,
  computeRemainingDistance,
  computeRemainingTime,
  determineNextInstruction,
  startNavigation,
} from '@/utils/navigationUtils';

export interface NavigationProps {
  navigationState: NavigationState;
  setNavigationState: React.Dispatch<SetStateAction<NavigationState>>;
  destination: { latitude: number; longitude: number };
  setDestination: React.Dispatch<SetStateAction<LatLng>>;
  travelMode: TravelMode;
  setTravelMode: React.Dispatch<SetStateAction<TravelMode>>;
  setSnappedPoint: React.Dispatch<SetStateAction<LatLng | null>>;
  clippedPolyline: LatLng[] | null;
  setClippedPolyline: React.Dispatch<SetStateAction<LatLng[] | null>>;
}

export const Navigation: React.FC<NavigationProps> = ({
  navigationState,
  setNavigationState,
  destination,
  setDestination,
  travelMode,
  setTravelMode,
  setSnappedPoint,
  clippedPolyline,
  setClippedPolyline,
}) => {
  const [is3d, setIs3d] = useState(true);
  const [routeData, setRouteData] = useState<RouteData | null>(null);
  const [remainingDistance, setRemainingDistance] = useState<number>(0);
  const [remainingTime, setRemainingTime] = useState<number>(0);
  const [nextInstruction, setNextInstruction] = useState<string>('');

  const { location } = useMapLocation();

  useEffect(() => {
    (async () => {
      if (!location) return;
      if (navigationState === 'default') return;
      try {
        const data = await getGoogleMapsRoute(
          location,
          destination,
          travelMode
        );
        setRouteData(data);
        {
          navigationState === 'planning' &&
            mapRef.current?.fitToCoordinates(data.polyline, {
              edgePadding: {
                top: 80,
                right: 60,
                bottom: 80,
                left: 60,
              },
              animated: true,
            });
        }
      } catch (error) {
        console.error('Failed to fetch route data:', error);
      }
    })();
  }, [destination, travelMode]);

  useEffect(() => {
    if (location && routeData) {
      const turfLine = lineString(
        routeData.polyline.map(coord => [coord.longitude, coord.latitude])
      );
      const userPoint = point([location.longitude, location.latitude]);
      const snapped = nearestPointOnLine(turfLine, userPoint);
      if (snapped?.geometry?.coordinates) {
        const [lng, lat] = snapped.geometry.coordinates;
        const snappedLoc: LatLng = { latitude: lat, longitude: lng };
        setSnappedPoint(snappedLoc);

        const newRemainingTime = computeRemainingTime(
          routeData.steps,
          snappedLoc,
          routeData.totalDuration
        );
        const newRemainingDistance = computeRemainingDistance(
          routeData.steps,
          snappedLoc,
          routeData.totalDistance
        );
        setRemainingTime(newRemainingTime);
        setRemainingDistance(newRemainingDistance);

        const instruction = determineNextInstruction(
          routeData.steps,
          snappedLoc,
          travelMode
        );
        setNextInstruction(instruction);

        const clipped = clipPolylineFromSnappedPoint(
          routeData.polyline,
          snappedLoc
        );
        setClippedPolyline(clipped);
      }
    }
  }, [location, routeData, travelMode]);

  useEffect(() => {
    if (
      navigationState === 'navigating' &&
      is3d &&
      location &&
      clippedPolyline
    ) {
      startNavigation(location, clippedPolyline, mapRef);
    }
  }, [location, navigationState]);

  useEffect(() => {
    const arrivalThreshold = 10; // meters
    if (
      navigationState === 'navigating' &&
      remainingDistance <= arrivalThreshold
    ) {
      bottomSheetRef.current?.snapToIndex(1);
      handleCurrentLocation(mapRef, location);
      setNavigationState('default');
    }
  }, [remainingDistance, navigationState]);

  const handleStartNavigation = () => {
    if (!location || !clippedPolyline) return;
    setNavigationState('navigating');
    startNavigation(location, clippedPolyline, mapRef);
  };

  return (
    <>
      {navigationState === 'default' && (
        <TouchableOpacity
          style={styles.navigateButton}
          onPress={() => {
            bottomSheetRef.current?.close();
            setNavigationState('planning');
            setDestination({
              latitude: 45.4808743,
              longitude: -73.6199895,
            });
          }}
        >
          <Text style={styles.text}>Navigation Demo</Text>
        </TouchableOpacity>
      )}

      {navigationState === 'planning' && (
        <TravelModeToggle
          selectedMode={travelMode}
          onModeSelect={setTravelMode}
        />
      )}

      {navigationState === 'navigating' && (
        <Instructions instruction={nextInstruction} />
      )}

      {(navigationState === 'planning' || navigationState === 'navigating') &&
        routeData && (
          <>
            <NavigationInfo
              duration={remainingTime}
              distance={remainingDistance}
              isNavigating={navigationState === 'navigating'}
              is3d={is3d}
              travelMode={travelMode}
              destination={destination}
              updateIs3d={setIs3d}
              onCancel={() => {
                bottomSheetRef.current?.snapToIndex(1);
                handleCurrentLocation(mapRef, location);
                setNavigationState('default');
              }}
              onStartNavigation={handleStartNavigation}
            />
          </>
        )}
    </>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
  map: {
    width: '100%',
    height: '100%',
  },
  navigateButton: {
    position: 'absolute',
    top: 70,
    left: 15,
    backgroundColor: 'rgba(34, 34, 34, 0.992)',
    borderRadius: 12,
    height: 36,
    width: 140,
    justifyContent: 'center',
    alignItems: 'center',
  },
  text: {
    color: 'white',
  },
});
