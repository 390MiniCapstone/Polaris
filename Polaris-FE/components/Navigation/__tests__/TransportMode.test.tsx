jest.mock('@expo/vector-icons', () => {
  const React = require('react');
  const { View } = require('react-native');
  const Icon = (props: any) => <View {...props} />;
  return {
    Ionicons: Icon,
    MaterialIcons: Icon,
    FontAwesome: Icon,
    FontAwesome5: Icon,
    FontAwesome6: Icon,
    MaterialCommunityIcons: Icon,
    Entypo: Icon,
    SimpleLineIcons: Icon,
  };
});

jest.mock('expo-font', () => ({
  loadAsync: jest.fn(),
  isLoaded: jest.fn(() => true),
}));

import React from 'react';
import { render, fireEvent } from '@testing-library/react-native';
import { StyleSheet } from 'react-native';
import { TravelModeToggle } from '@/components/Navigation/TravelModeToggle';

const modes: Array<'DRIVE' | 'WALK' | 'TRANSIT' | 'BICYCLE'> = [
  'DRIVE',
  'WALK',
  'TRANSIT',
  'BICYCLE',
];

describe('TravelModeToggle Component', () => {
  it('renders correctly (snapshot)', () => {
    const { toJSON } = render(
      <TravelModeToggle selectedMode="DRIVE" onModeSelect={jest.fn()} />
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('calls onModeSelect with the correct mode when each button is pressed', () => {
    const onModeSelectMock = jest.fn();
    const { getByTestId } = render(
      <TravelModeToggle selectedMode="DRIVE" onModeSelect={onModeSelectMock} />
    );

    modes.forEach(mode => {
      const button = getByTestId(`transport-mode-button-${mode}`);
      fireEvent.press(button);
      expect(onModeSelectMock).toHaveBeenCalledWith(mode);
    });
    expect(onModeSelectMock).toHaveBeenCalledTimes(modes.length);
  });

  it('applies the selected style to the button corresponding to the selected mode', () => {
    const { getByTestId } = render(
      <TravelModeToggle selectedMode="TRANSIT" onModeSelect={jest.fn()} />
    );

    const transitButton = getByTestId('transport-mode-button-TRANSIT');
    const transitStyle = StyleSheet.flatten(transitButton.props.style);

    expect(transitStyle.backgroundColor).toBe('#9A2E3F');
  });
});
