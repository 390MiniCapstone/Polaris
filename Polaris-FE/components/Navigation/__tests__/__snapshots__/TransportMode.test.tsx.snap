// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TravelModeToggle Component renders correctly (snapshot) 1`] = `
<View
  style={
    {
      "flexDirection": "row",
      "gap": 14,
      "justifyContent": "center",
      "left": 0,
      "position": "absolute",
      "right": 0,
      "top": 76,
    }
  }
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#9A2E3F",
        "borderRadius": 15,
        "height": 40,
        "justifyContent": "center",
        "opacity": 1,
        "width": 60,
      }
    }
    testID="transport-mode-button-DRIVE"
  >
    <View
      color="white"
      name="car"
      size={22}
    />
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "rgba(34, 34, 34, 0.992)",
        "borderRadius": 15,
        "height": 40,
        "justifyContent": "center",
        "opacity": 1,
        "width": 60,
      }
    }
    testID="transport-mode-button-WALK"
  >
    <View
      color="white"
      name="walking"
      size={22}
    />
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "rgba(34, 34, 34, 0.992)",
        "borderRadius": 15,
        "height": 40,
        "justifyContent": "center",
        "opacity": 1,
        "width": 60,
      }
    }
    testID="transport-mode-button-TRANSIT"
  >
    <View
      color="white"
      name="bus-simple"
      size={22}
    />
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "rgba(34, 34, 34, 0.992)",
        "borderRadius": 15,
        "height": 40,
        "justifyContent": "center",
        "opacity": 1,
        "width": 60,
      }
    }
    testID="transport-mode-button-BICYCLE"
  >
    <View
      color="white"
      name="bicycle"
      size={22}
    />
  </View>
</View>
`;
